city_metrix/layers/ndvi_sentinel2_gee.py

```diff
@@ -0,0 +1,65 @@
+import ee
+from tools.xarray_tools import convert_ratio_to_percentage
+from .layer import Layer, get_image_collection
+
+class NdviSentinel2(Layer):
+    """"
+    NDVI = Sentinel-2 Normalized Difference Vegetation Index
+    param: year: The satellite imaging year.
+    return: a rioxarray-format DataArray
+    Author of associated Jupyter notebook: Ted.Wong@wri.org
+    Notebook: https://github.com/wri/cities-cities4forests-indicators/blob/dev-eric/scripts/extract-VegetationCover.ipynb
+    Reference: https://en.wikipedia.org/wiki/Normalized_difference_vegetation_index
+    """
+    def __init__(self, year=None, **kwargs):
+        super().__init__(**kwargs)
+        self.year = year
+
+    def get_data(self, bbox):
+        if self.year is None:
+            raise Exception('NdviSentinel2.get_data() requires a year value')
+
+        start_date = "%s-01-01" % self.year
+        end_date = "%s-12-31" % self.year
+
+        # Compute NDVI for each image
+        def calculate_ndvi(image):
+            ndvi = (image
+                    .normalizedDifference(['B8', 'B4'])
+                    .rename('NDVI'))
+            return image.addBands(ndvi)
+
+        s2 = ee.ImageCollection("COPERNICUS/S2_HARMONIZED")
+        ndvi = (s2
+                .filterBounds(ee.Geometry.BBox(*bbox))
+                .filterDate(start_date, end_date)
+                .map(calculate_ndvi)
+                .select('NDVI')
+                 )
+
+        ndvi_mosaic = ndvi.qualityMosaic('NDVI')
+
+        ic = ee.ImageCollection(ndvi_mosaic)
+        ndvi_data = get_image_collection(ic, bbox, 10, "NDVI")
+
+        xdata = ndvi_data.to_dataarray()
+
+        return xdata
+
+    def prepare_for_map_rendering(self, data, ndvi_threshold=0.4, convert_to_percentage=True, **kwargs):
```

Comment:
The only thing I'm unsure about here is if these steps are really just for map rendering, or they're just an alternative use of the data? I think the threshold is for the indicator/metric, but I know certain users are just interested in the actual NDVI values so they can calculate different metrics.

So I don't think this is specifically related to map rendering. Maybe instead there should be a subclass of NDVI that accepts a threshold, and applies it at the end?

Reply:
I debated whether to make it a generic or specific term. Examples of generic were simply "postprocess" or "reduce_data_volume" to indicate the functionality. I currently opted for a specific term describing the intended specific application (as I understand it). 

I add this as an abstract method on the Layer class, so it would be good to have a name that has broad applicability. 

We could have other post-processing functions or include more and alternative controls on post-processing of other layers.

@tedw0ng - Two questions: 1) Do you agree with the approach of returning the raw NDVI value by default from get_data() and have secondary processing for modifying it? 2) Thinking broadly across how this post-processing could be used, what would you recommend for a function name?

Reply:
Post-processing method removed from Layers.